### PR TITLE
Build applications based on ApplicationConfig

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: examples/hello_world/module/cpp/hello_world.cc [hunk 1]
```diff
@@ -37,29 +37,39 @@ enum OakStatus {
...
 
 WASM_IMPORT("oak") int wait_on_channels(uint8_t* buff, int32_t count);
 WASM_IMPORT("oak")
 int channel_read(uint64_t handle, uint8_t* buff, size_t usize, uint32_t* actual_size);
 WASM_IMPORT("oak") int channel_write(uint64_t handle, uint8_t* buff, size_t usize);
+WASM_IMPORT("oak") uint64_t channel_find(uint8_t* buff, size_t usize);
 
 WASM_EXPORT int32_t oak_main() {
-  uint8_t handle_space[9] = {oak::ChannelHandle::GRPC_IN, 0, 0, 0, 0, 0, 0, 0, 0};
+  char grpc_in_name[] = "grpc_in";
```
Comment (tiziano88) [inline examples/hello_world/module/cpp/hello_world.cc]: Why not `std::string`?

File: oak/proto/oak_api.proto [hunk 2]
```diff
@@ -35,14 +35,4 @@ enum OakStatus {
...
-// Well-known channel handle values.
-enum ChannelHandle {
-  CHANNEL_HANDLE_UNSPECIFIED = 0;
-  LOGGING = 1;   // Write half of channel
-  GRPC_IN = 2;   // Read half of channel
-  GRPC_OUT = 3;  // Write half of channel
-  STORAGE_IN = 4;
-  STORAGE_OUT = 5;
-}
+}
```
Comment (tiziano88) [inline oak/proto/oak_api.proto]: (nit) missing newline at EOF

File: oak/server/oak_grpc_node.h [hunk 3]
```diff
@@ -29,12 +29,14 @@ class OakGrpcNode final : public Application::Service {
   grpc::Status Start();
   grpc::Status Stop();
 
-  void SetUpGrpcChannels(std::shared_ptr<MessageChannel> grpc_in,
-                         std::shared_ptr<MessageChannel> grpc_out);
+  // Add the inbound channel (with implicit port name "response").
+  void AddReadChannel(std::unique_ptr<MessageChannelReadHalf> rsp_half);
```
Comment (tiziano88) [inline oak/server/oak_grpc_node.h]: (suggestion) `SetResponseChannel`, and similarly elsewhere (even `LoggingNode::SetLogChannel`).

File: oak/server/oak_node.h [hunk 4]
```diff
@@ -31,19 +31,21 @@
...
 namespace oak {
 
-typedef std::unordered_map<Handle, std::unique_ptr<ChannelHalf>> ChannelHalfTable;
+using ChannelHalfTable = std::unordered_map<Handle, std::unique_ptr<ChannelHalf>>;
 
 class OakNode final : public NodeThread {
  public:
   // Creates an Oak node by loading the Wasm module code.
-  static std::unique_ptr<OakNode> Create(const std::string& module);
+  static std::unique_ptr<OakNode> Create(const std::string& name, const std::string& module);
```
Comment (tiziano88) [inline oak/server/oak_node.h]: I assume the strings pointed to by `name` and `module` need to outlive this class?

File: oak/server/oak_node.cc [hunk 5]
```diff
@@ -359,4 +369,25 @@ wabt::interp::HostFunc::Callback OakNode::OakWaitOnChannels(wabt::interp::Enviro
...
+  return [this, env](const wabt::interp::HostFunc* func, const wabt::interp::FuncSignature* sig,
+                     const wabt::interp::TypedValues& args, wabt::interp::TypedValues& results) {
+    LogHostFunctionCall(func, args);
+
+    uint32_t offset = args[0].get_i32();
+    uint32_t length = args[1].get_i32();
+    results[0].set_i64(0);
+
+    auto base = env->GetMemory(0)->data.begin() + offset;
+    std::string port_name(base, base + length);
```
Comment (tiziano88) [inline oak/server/oak_node.cc]: Do we need to check boundaries here? AFAICT this may well be past the end of the memory.

File: oak/server/oak_runtime.cc [hunk 6]
```diff
@@ -68,81 +112,36 @@ asylo::Status OakRuntime::Start() {
...
+    logging_node_->Start();
+  }
 
   // Start a new thread to process storage requests.
-  storage_node_->Start();
+  if (storage_node_ != nullptr) {
+    storage_node_->Start();
+  }
 
   // Now all dependencies are running, start the thread for the Node itself.
```
Comment (tiziano88) [inline oak/server/oak_runtime.cc]: Reword this to account for multiple nodes.

File: README.md [hunk 7]
```diff
@@ -119,24 +119,26 @@ messages, with a receive half and a send half that an Oak module can read from
 or write to respectively. Each half of a channel is identified by a **handle**,
 which is used as a parameter to the corresponding host function calls.
 
-At each invocation, the following channel halves are implicitly available to the Oak
-Node (see [/oak/common/handles.h](oak/common/handles.h)):
+A collection of pre-configured channel halves are available to the Oak Node.
```
Comment (tiziano88) [inline README.md]: I don't think these channels should be pre-configured at all? It's up to the application configuration to specify port names for the Wasm node, so if that specifies a different port name, that's the one that will be available from the node, right?

File: oak/server/oak_runtime.cc [hunk 8]
```diff
@@ -35,31 +35,75 @@ asylo::Status OakRuntime::Initialize(const ApplicationConfiguration& config) {
...
+    const std::string& dest_node = channel_config.destination_endpoint().node_name();
+    const std::string& dest_port = channel_config.destination_endpoint().port_name();
+
+    LOG(INFO) << "Create channel " << src_node << "." << src_port << " -> " << dest_node << "."
+              << dest_port;
+    auto channel = std::make_shared<MessageChannel>();
+    auto write_half = absl::make_unique<MessageChannelWriteHalf>(channel);
+    auto read_half = absl::make_unique<MessageChannelReadHalf>(channel);
+
+    // Wire up the source endpoint.
```
Comment (tiziano88) [inline oak/server/oak_runtime.cc]: This seems a very manual wiring, in general I would think all nodes should just be indexed in a map, and we just look up nodes by name and associate the channel to the port defined in the application configuration via `AddNamedChannel`.

File: oak/server/oak_runtime.cc [hunk 9]
```diff
@@ -68,81 +112,36 @@ asylo::Status OakRuntime::Start() {
   LOG(INFO) << "Starting runtime";
 
   // Start the gRPC pseudo-node.
-  grpc_node_->Start();
+  if (grpc_node_ != nullptr) {
```
Comment (tiziano88) [inline oak/server/oak_runtime.cc]: This also seems very manual, we should just put all the nodes (and pseudo-nodes) in a map, iterate and start all of them. WDYT?

File: oak/proto/oak_api.proto (hunk 2 above)
Comment (daviddrysdale) [reply oak/proto/oak_api.proto]: Done

File: README.md (hunk 7 above)
Comment (daviddrysdale) [reply README.md]: I've improved the text a bit for this, but I don't want to get into fully documenting the `ApplicationConfig` stuff yet.

File: examples/hello_world/module/cpp/hello_world.cc (hunk 1 above)
Comment (daviddrysdale) [reply examples/hello_world/module/cpp/hello_world.cc]: The Wasm C++ build isn't set up to include libstdc++ stuff yet AFACT.

File: oak/server/oak_node.h (hunk 4 above)
Comment (daviddrysdale) [reply oak/server/oak_node.h]: I don't think so &ndash; `name` gets copy-constructed into a new `std::string` and `module` ends up at `wabt::ReadBinaryInterp()` which (I think) copies out the relevant sections.

File: oak/server/oak_node.cc (hunk 5 above)
Comment (daviddrysdale) [reply oak/server/oak_node.cc]: Pre-existing so added #219

File: oak/server/oak_runtime.cc (hunk 8 above)
Comment (daviddrysdale) [reply oak/server/oak_runtime.cc]: Yeah, I'm planning to homogenize nodes & pseudo-nodes later.  Added a TODO for now.

File: oak/server/oak_runtime.cc (hunk 6 above)
Comment (daviddrysdale) [reply oak/server/oak_runtime.cc]: Done.

File: oak/server/oak_runtime.cc (hunk 9 above)
Comment (daviddrysdale) [reply oak/server/oak_runtime.cc]: Yep, planning to do something along these lines later on.

File: oak/server/oak_grpc_node.h (hunk 3 above)
Comment (daviddrysdale) [reply oak/server/oak_grpc_node.h]: These were named for similarity with `AddNamedChannel`, and will hopefully get removed in favour of that in a later set of commits.
